spatialmath/capsule.go

```diff
@@ -0,0 +1,346 @@
+func (c *capsule) AlmostEqual(g Geometry) bool {
+	other, ok := g.(*capsule)
+	if !ok {
+		return false
+	}
+	return PoseAlmostEqual(c.pose, other.pose) &&
+		utils.Float64AlmostEqual(c.radius, other.radius, 1e-8) &&
+		utils.Float64AlmostEqual(c.length, other.length, 1e-8)
+}
+
+// Transform premultiplies the capsule pose with a transform, allowing the capsule to be moved in space.
+func (c *capsule) Transform(toPremultiply Pose) Geometry {
+	return newCapsuleWithSegPoints(Compose(toPremultiply, c.pose), c.radius, c.length, c.label)
+}
+
+// ToProto converts the capsule to a Geometry proto message.
+func (c *capsule) ToProtobuf() *commonpb.Geometry {
+	return &commonpb.Geometry{
+		Center: PoseToProtobuf(c.pose),
+		GeometryType: &commonpb.Geometry_Capsule{
+			Capsule: &commonpb.Capsule{
+				RadiusMm: c.radius,
+				LengthMm: c.length,
+			},
+		},
+		Label: c.label,
+	}
+}
+
+// CollidesWith checks if the given capsule collides with the given geometry and returns true if it does.
+func (c *capsule) CollidesWith(g Geometry) (bool, error) {
+	if other, ok := g.(*box); ok {
+		return capsuleVsBoxCollision(c, other), nil
+	}
+	dist, err := c.DistanceFrom(g)
+	if err != nil {
+		return true, err
+	}
+	return dist <= CollisionBuffer, nil
+}
+
+// CollidesWith checks if the given capsule collides with the given geometry and returns true if it does.
+func (c *capsule) DistanceFrom(g Geometry) (float64, error) {
+	if other, ok := g.(*box); ok {
+		return capsuleVsBoxDistance(c, other), nil
+	}
+	if other, ok := g.(*capsule); ok {
+		return capsuleVsCapsuleDistance(c, other), nil
+	}
+	if other, ok := g.(*point); ok {
+		return capsuleVsPointDistance(c, other.position), nil
+	}
+	if other, ok := g.(*sphere); ok {
+		return capsuleVsSphereDistance(c, other), nil
+	}
+	return math.Inf(-1), newCollisionTypeUnsupportedError(c, g)
+}
+
+func (c *capsule) EncompassedBy(g Geometry) (bool, error) {
+	if other, ok := g.(*capsule); ok {
+		return capsuleInCapsule(c, other), nil
+	}
+	if other, ok := g.(*box); ok {
+		return capsuleInBox(c, other), nil
+	}
+	if other, ok := g.(*sphere); ok {
+		return capsuleInSphere(c, other), nil
+	}
+	if _, ok := g.(*point); ok {
+		return false, nil
+	}
+	return true, newCollisionTypeUnsupportedError(c, g)
+}
+
+// ToPoints converts a capsule geometry into []r3.Vector. This method takes one argument which determines
+// how many points should like on the capsule's surface. If the argument is set to 0. we automatically
+// substitute the value with defaultTotalSpherePoints.
+func (c *capsule) ToPoints(resolution float64) []r3.Vector {
+	s := &sphere{pose: NewZeroPose(), radius: c.radius}
+	vecList := s.ToPoints(resolution)
+	// move points to be correctly located on capsule endcaps
+	adj := c.length/2 - c.radius
+	for _, pt := range vecList {
+		if pt.Z >= 0 {
+			pt.Z += adj
+		} else {
+			pt.Z -= adj
+		}
+	}
+
+	// Now distribute points along the cylindrical shaft
+	totalShaftPts := (c.radius * c.length) * resolution
```

Comment:
It doesn't look like we're using `defaultTotalSpherePoints` here (which is a little strange name-wise), but I think we need something since we could then have a `totalShaftPoints` of 0 with default sphere point densities on each end of the capsule. Does that make sense?

Reply:
Fixed to ensure resolution is not zero